staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -452,15 +578,27 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
 	// this could happen if the create event is merged from create-update events
 	storageMap := r.customStorage.Load().(crdStorageMap)
 	oldInfo, found := storageMap[crd.UID]
-	if !found {
-		return
-	}
-	if apiequality.Semantic.DeepEqual(&crd.Spec, oldInfo.spec) && apiequality.Semantic.DeepEqual(&crd.Status.AcceptedNames, oldInfo.acceptedNames) {
-		klog.V(6).Infof("Ignoring customresourcedefinition %s create event because a storage with the same spec and accepted names exists",
-			crd.Name)
-		return
+
+	var tearDownFinishedCh <-chan struct{}
+	if found {
+		if apiequality.Semantic.DeepEqual(&crd.Spec, oldInfo.spec) && apiequality.Semantic.DeepEqual(&crd.Status.AcceptedNames, oldInfo.acceptedNames) {
+			klog.V(6).Infof("Ignoring customresourcedefinition %s create event because a storage with the same spec and accepted names exists",
+				crd.Name)
+			return
+		}
+		// Update storage version with the latest info in the watch event
+		tearDownFinishedCh = r.removeStorageLocked(crd.UID)
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		ctx := apirequest.NewContext()
+		// customStorageLock will be released even if UpdateStorageVersion() fails. This is safe
+		// since we are deleting the old storage here and not creating a new one.
+		err := r.storageVersionManager.UpdateStorageVersion(ctx, crd, tearDownFinishedCh, nil, nil)
```

Comment:
This call turns this method from non-blocking to blocking since the `UpdateStorageVersion` won't complete until `tearDownFinishedCh` is closed.  The async looks worthwhile to avoid blocking and to make the code more responsive to connections that haven't closed.

Reply:
So thinking through this and the `UpdateStorageVersion` call in in `getOrCreateServingInfoFor`.

If we blindly make this async to unblock handling of other CRDs (probably required), what could happen.

1. standing CRD storage version is v1
2. CRD is updated with storage version v2
3. that update is observed and this cleanup method is called
4. a CR CREATE is in progress and extremely slow, it will store in v1.  This (properly) stops the `UpdateStorageVersion` from being called until the requests are drained from the old handler
5. a new CR request is made to create a new instance
6. `getOrCreateServingInfoFor` has a call to `UpdateStorageVersion` which does not wait (no wait channel passed)
7. storageversion is updated to indicate v2 instead of v1
8. very quickly we do a storage version migration because it's now safe and it completes
9. the CR CREATE request from step 4 finishes, result is an instance stored in v1
10. we now have data stored in v1 format in etcd. This isn't expected.


Trying to think through what conditions need to be correct in order to change the storage version, but I think this is racy.

Reply:
I see the potential for a resource in a long-running request to be written in an older storage version if we naively make the update async.


Doesn't the solution as proposed still leave the possibility a write to complete in a version that is not the currently active StorageVersion? 

I suppose the distinction is if you are stepping the crd incrementally (not going backwards) any new resources or updates should always be stored at a version at least as high as the published storage version.? That seems OK for SVM's use case. Is this a guarantee for this Storage Version API?

Reply:
> I suppose the distinction is if you are stepping the crd incrementally (not going backwards) any new resources or updates should always be stored at a version at least as high as the published storage version.?

Yes, that's the distinction that I see.

> Is this a guarantee for this Storage Version API?

The guarantee for in-tree resources is stronger. I think to achieve better, we would need to block writes to the CR until all previous storage has been destroyed.  This could be built and may not be too bad to construct.